Support classExpressions in DTE

h-joo opened the pull request:
Except for the commit e9ca3d4d36924a880dda29ed3e1f0348f03ca092 everything else is non-functional, all others are fixing tests and moving things around, formatting.

h-joo committed:
Move ObjectLiteral handling into a function

This is a non-functional move.

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Remove the function mergeFlags

as it's doing trivial bitwise OR operation, also some other
parts of the code is doing |= instead, so try to be consistent.

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Do not attempt to generate d.ts files

if there's any error reported.

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Improve type aesthetics

to reduce unncessary type checks

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Merge single use function into it's caller

Also, avoid unncessary work caused by calling invalid(),
rather call reportIsolatedDeclarationError instead as it was the only
thing intended.

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Be explicit of the flag AND operation

to see where it's actually being used.

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Tidy up the public interfaces and accept baseline

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Fix test cases which were failing due to typos

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Update baselines

These seems all reasonable, we might have missed
something before that was not updated.

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Support class expressions in DTE

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Format files with hereby format

This commit is totally non-functional

Signed-off-by: Hana Joo <hanajoo@google.com>

dragomirtitian commented on src/compiler/transformers/declarations.ts:
```diff
@@ -331,7 +331,7 @@ export function transformDeclarations(context: TransformationContext) {
     let emittedImports: readonly AnyImportSyntax[] | undefined; // must be declared in container so it can be `undefined` while transformer's first pass
     const resolver = context.getEmitResolver();
-    const localInferenceResolver = createLocalInferenceResolver({
+    const { resolver: localInferenceResolver, isolatedDeclarations } = createLocalInferenceResolver({
```
Like the discriminated union destructuring here.

dragomirtitian commented on src/compiler/transformers/declarations/localInferenceResolver.ts:
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
+        const constructorParameters: ParameterDeclaration[] = [];
+
+        if (node.heritageClauses && node.heritageClauses.length > 0) {
+            context.addDiagnostic({
+                ...createDiagnosticForNode(node, Diagnostics.Declaration_emit_for_this_file_requires_type_resolution_An_explicit_type_annotation_may_unblock_declaration_emit),
+                relatedInformation: [
+                    createDiagnosticForRange(
+                        currentSourceFile,
+                        {
+                            pos: node.heritageClauses[0].pos,
+                            end: node.heritageClauses[node.heritageClauses.length - 1].end,
+                        },
+                        Diagnostics.Heritage_clauses_in_class_expressions_are_not_allowed_with_isolatedDeclarations,
+                    ),
+                ],
+            });
+            invalid = true;
```
Really sad we have to not allow heritage clauses. 

I think we need to rephrase this a bit "not allowed" implies to me that you can't use them at all. You can, provided you add an annotation to the variable, we just can't infer from them.

h-joo replied:
I made the message : "To use heritage clauses in class expressions with --isolatedDeclarations, you need explicit type annotation on the variable."

dragomirtitian commented on src/compiler/transformers/declarations/localInferenceResolver.ts:
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
+                else {
+                    nonStaticMembers.push(methodSignature);
+                }
+            }
```
This seems very similar to inference for methods on object literals. Only it does not do inference for parameters with default values. Can we factor out the code in the object inference and reuse it here?

h-joo replied:
Done

dragomirtitian commented on tests/baselines/reference/esDecorators-classDeclaration-setFunctionName(target=es2022).js:
```diff
@@ -62,20 +62,24 @@ let C = (() => {
 //// [c.js]
-export default (() => {
+let C = (() => {
     let _classDecorators = [dec];
     let _classDescriptor;
     let _classExtraInitializers = [];
     let _classThis;
     var C = class {
+        static { _classThis = this; }
         static {
-            __esDecorate(null, _classDescriptor = { value: this }, _classDecorators, { kind: "class", name: this.name }, null, _classExtraInitializers);
+            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
+            __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
             C = _classThis = _classDescriptor.value;
+            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
             __runInitializers(_classThis, _classExtraInitializers);
         }
     };
     return C = _classThis;
 })();
+export default C;
```
Do we know why this changed ?

h-joo replied:
I have no idea, this seemed to be broken already in the target branch, but I don't have a good imagination to understand what would have caused this to change.

dragomirtitian commented on src/compiler/transformers/declarations/localInferenceResolver.ts:
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
+                }
+            }
+            else if (isGetAccessorDeclaration(member) || isSetAccessorDeclaration(member)) {
+                if (!hasGetSetAccessor) {
+                    hasGetSetAccessor = true;
+                    let type;
+                    if (isGetAccessorDeclaration(member)) {
+                        type = localInferenceFromInitializer(member, member.type);
+                    }
+                    else {
+                        type = localInferenceFromInitializer(member.parameters[0], member.parameters[0].type);
+                    }
+                    if (!type) {
+                        invalid = true;
+                    }
+                    nonStaticMembers.push(
+                        factory.createPropertySignature(
+                            [],
+                            member.name,
+                            /*questionToken*/ undefined,
+                            type,
+                        ),
+                    );
+                }
```
Similar to the object inference maybe we can refactor to a function. Also it should infer setter/getter type from the other accessor if present,

h-joo replied:
Done.

dragomirtitian commented on src/compiler/transformers/declarations/localInferenceResolver.ts:
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
+                }
+            }
+            else if (isGetAccessorDeclaration(member) || isSetAccessorDeclaration(member)) {
+                if (!hasGetSetAccessor) {
+                    hasGetSetAccessor = true;
```
Not sure I understand this. What if the class has multiple getter/setters for different names

h-joo replied:
you're right, I don't know what I was thinking while writing this code. I've adopted your code anyways so no longer relevant.

dragomirtitian commented on src/compiler/transformers/declarations/localInferenceResolver.ts:
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
+                nonStaticMembers.push(member);
+            }
+            else if (isPropertyDeclaration(member)) {
+                const name = isPrivateIdentifier(member.name) ?
+                    // imitating the behavior from utilities.ts : getSymbolNameForPrivateIdentifier, but as we don't have
+                    // a Symbol & SymbolId in hand, we use NodeId of the declaration instead as an approximiation and to provide uniqueness.
+                    // TODO: This seems to have a high collision possibilitiy than the vanilla implementation as we have much less
+                    // ids for nodes in DTE.
+                    factory.createStringLiteral(`__#${node.parent.id}@${member.name.escapedText}`) :
+                    member.name;
```
I think the appropriate question here is *why* does TS keep private names in declarations at all. This feels like a TS bug. For class declarations the members are removed and `#private` is added to the class. There is no reason to expose what private members the class expression has and the type they have. 

```tsx
const x = class {
    #id = 10 // "__#3455@#id": number; in dedclarations 
}
class X {
    #id = 10 // removed and #private; added
}
```

[Playground Link](https://www.typescriptlang.org/play?ts=5.3.0-beta#code/MYewdgzgLgBAHjAvDYAbAhhCMDeBYAKBmJgGIBLAEyRgEYAGGAeiZgCIB9D0gZgBYArAIACFSmwBcMMAFcAtgCMApgCcA3DHJgYlJZTToV6KOXDZCAX0IGsMABq5CJMlRoNmrFUrkgAbnph0MGpSAAcVcl9jJQ10Sl1KSyA)

h-joo replied:
Okay, I'll put into the agenda for next week's meeting.

dragomirtitian commented on src/compiler/transformers/declarations/localInferenceResolver.ts:
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
+                    // ids for nodes in DTE.
+                    factory.createStringLiteral(`__#${node.parent.id}@${member.name.escapedText}`) :
+                    member.name;
+                const type = localInferenceFromInitializer(member, member.type);
+                if (!type) {
+                    invalid = true;
+                }
+                const propertySignature = factory.createPropertySignature(
+                    keepReadonlyKeyword(member.modifiers),
+                    name,
+                    member.questionToken,
+                    type,
+                );
+                if (member.modifiers?.some(modifier => modifier.kind === SyntaxKind.StaticKeyword)) {
+                    staticMembers.push(propertySignature);
+                }
+                else {
+                    nonStaticMembers.push(propertySignature);
+                }
```
Similar to Property assignment from object literals, again maybe we can reuse ?

h-joo replied:
Done.

h-joo committed:
Fix PR Comments

1. Reuse reusable chunk of code.
2. Proper handling of accessors, before it treated as if there were only
   a single accessor in the class.
3. Improve wording in the error message for heritage clauses

Signed-off-by: Hana Joo <hanajoo@google.com>

dragomirtitian commented on src/compiler/transformers/declarations/localInferenceResolver.ts:
```diff
@@ -388,7 +398,7 @@ export function createLocalInferenceResolver({
             if (isComputedPropertyName(prop.name)) {
                 if (!resolver.isLiteralComputedName(prop.name)) {
-                    reportIsolatedDeclarationError(node);
+                    reportIsolatedDeclarationError(objectLiteral);
```
I think this is wrong, we should report on `prop.name` to help the user identify the issue. I know this was the case before, but could you fix it as long as we are making changes here?

h-joo replied:
Done

dragomirtitian commented on src/compiler/transformers/declarations/localInferenceResolver.ts:
```diff
@@ -556,46 +583,26 @@ export function createLocalInferenceResolver({
-                const type = localInferenceFromInitializer(member, member.type);
-                if (!type) {
+                const { method, flags } = handleMethodDeclaration(member, member.name, NarrowBehavior.None);
+                if (flags === LocalTypeInfoFlags.Invalid) {
```
Could you please test with `&` for the invalid flag. Currently there is only one flag (used to be more) but I would like to keep the idea of this being a flag enum at least for now.

h-joo replied:
Done

h-joo committed:
Fix more PR Comments

1. use bitwise AND instead of comparison
2. report errors on property name for object literals instead of the
   object literal

Signed-off-by: Hana Joo <hanajoo@google.com>
